src/attach-shadow.js

```diff
@@ -92,94 +96,145 @@ ShadyRoot.prototype._rendererForHost = function() {
       c = c$[i];
-      if (this._distributor.isInsertionPoint(c)) {
+      if (this._isInsertionPoint(c)) {
         return root;
       }
     }
   }
 }
 
-ShadyRoot.prototype.render = function() {
+ShadyRoot.prototype._render = function() {
   if (this._renderPending) {
-    this._getRenderRoot()['_render']();
+    this._getRenderRoot()['_renderRoot']();
   }
 }
 
 // NOTE: avoid renaming to ease testability.
-ShadyRoot.prototype['_render'] = function() {
+ShadyRoot.prototype['_renderRoot'] = function() {
   this._renderPending = false;
-  this._changePending = false;
-  if (!this._skipUpdateInsertionPoints) {
-    this.updateInsertionPoints();
-  } else if (!this._hasRendered) {
-    this.__insertionPoints = [];
-  }
-  this._skipUpdateInsertionPoints = false;
-  // TODO(sorvell): can add a first render optimization here
-  // to use if there are no insertion points
-  // 1. clear host node of composed children
-  // 2. appendChild the shadowRoot itself or (more robust) its logical children
-  // NOTE: this didn't seem worth it in perf testing
-  // but not ready to delete this info.
-  // logical
-  this.distribute();
-  // physical
-  this.compose();
+  this._distribute();
+  this._compose();
   this._hasRendered = true;
 }
 
-ShadyRoot.prototype.forceRender = function() {
-  this._renderPending = true;
-  this.render();
+ShadyRoot.prototype._distribute = function() {
+  // capture # of previously assigned nodes to help determin if dirty.
+  for (let i=0, slot; i < this._slotList.length; i++) {
+    slot = this._slotList[i];
+    this._clearSlotAssignedNodes(slot);
+  }
+  if (!this._slotMap) {
+    this._updateSlotMap();
+  }
+  // distribute host children.
+  for (let n=this.host.firstChild; n; n=n.nextSibling) {
+    this._distributeNodeToSlot(n);
+  }
+  // fallback content, slotchange, and dirty roots
+  for (let i=0, slot; i < this._slotList.length; i++) {
+    slot = this._slotList[i];
+    // distribute fallback content
+    if (!slot.__shady.assignedNodes.length) {
+      for (let n=slot.firstChild; n; n=n.nextSibling) {
+        this._distributeNodeToSlot(n, slot);
+      }
+    }
+    let slotParentRoot = slot.parentNode.shadowRoot;
+    if (slotParentRoot && slotParentRoot._hasInsertionPoint()) {
+      slotParentRoot['_renderRoot']();
+    }
+    slot.__shady.distributedNodes = [];
+    this._addDistributedNodes(slot.__shady.distributedNodes, slot);
+    let prevAssignedNodes = slot.__shady._previouslyAssignedNodes;
+    if (prevAssignedNodes) {
+      for (let i=0; i < prevAssignedNodes.length; i++) {
+        prevAssignedNodes[i].__shady._prevAssignedSlot = null;
+      }
+      slot.__shady._previouslyAssignedNodes = null;
+      // dirty if previously less assigned nodes than previously assigned.
+      if (prevAssignedNodes.length > slot.__shady.assignedNodes.length) {
+        slot.__shady.dirty = true;
+      }
+    }
+    // NOTE: cannot bubble correctly here so not setting bubbles: true
```

Comment:
This note also appears in `_fireSlotChange`. It probably only needs to be in one place.